### DISCUSSION TITLE
What would a self hosted Production setup look like?

### DISCUSSION THREAD

**Discussion by traveltek-tmalek** [saleor/saleor#11877] [Q&A] [answered]
Hello. I've been asked to look into the pros and cons of self hosting Saleor in the Cloud (AWS) Vs. using the Cloud product.

I understand the 3 different tiers available in the cloud, but I am trying to calculate how much it would cost to run Saleor in AWS but there's something confusing me a little bit.

I also understand that Saleor is free and our only costs are going to be running Saleor on EC2 instances (or containers if we use AWS ECS).

Would we need 3 instances/3 containers. One for the core, one for the Storefront, and one for the dashboard? Or would we be installing the entire application on one instance (or container if going down that route).

This also makes me wonder about Specs. I read that 8 GB should be good for Saleor, but I am not sure if that applies to if you're running all 3 components of Saleor on one instance, or if each of the instances (x3) would need 8 GB of RAM.

Can anyone that has Saleor running in AWS (or a similar Cloud platform) share with me their setup so I get a better understanding and can start making cost calculations?

**Comment by traveltek-tmalek** [comment]
anyone?

**Comment by timuric** [comment, marked as answer]
Hey @traveltek-tmalek 

> I also understand that Saleor is free and our only costs are going to be running Saleor on EC2 instances (or containers if we use AWS ECS).

Yes, you would need to only pay for the infrastructure and dedicate time for maintenance.

> Would we need 3 instances/3 containers. One for the core, one for the Storefront, and one for the dashboard? Or would we be installing the entire application on one instance (or container if going down that route).

You would need only one container, here is breakdown of all components: 

1. Saleor is headless, therefore storefront can be hosted however you prefer, it can be SPA or Next.js hosted on platform such as Vercel or Netlify. Our Core APIs are backward compatible, so you can deploy Storefront updates independently. In a nutshell, it is entirely up to you, no matter if you are self-hosting or using cloud.
2. Dashboard doesn't need servers as it is an SPA, you only need to serve static files. Ideally, you would need to upgrade dashboard together with backend updates to provide feature parity.
3. Core/Backend would need a server, to provide seamless upgrades and auto-scale you would need container orchestration like Kubernetes.

> This also makes me wonder about Specs. I read that 8 GB should be good for Saleor, but I am not sure if that applies to if you're running all 3 components of Saleor on one instance, or if each of the instances (x3) would need 8 GB of RAM.

As mentioned above, you only need one instance to run Saleor Core as dashboard doesn't require servers. Configuration of server varies greatly on your usage patterns, amount of parallel requests, complexity of requests, database size. If you need auto-scaling that requires spontaneous spikes, you would need to allocate resources for that as well.
The best way to figure out is to run simulations and stress tests according to your business needs.

> I understand the 3 different tiers available in the cloud, but I am trying to calculate how much it would cost to run Saleor in AWS but there's something confusing me a little bit
There are scenarios where self-hosting might be cheaper and vice versa.

Scenarios when self-hosting might be cheaper:
• You have low average order value or small margins, therefore paying fee for orders is not ideal
• Low traffic
• High conversion rate, meaning most visitors purchase (for example b2b)
• Traffic is predictable, you don't have to invest into autoscaling

Scenarios when Cloud might be cheaper:
• High traffic of people browsing, therefore lower ratio of transactions to API calls
• Traffic spikes
• You need development environments and staging
• You wish to receive security patches and without maintenance

Overall, most clients that migrated from self-hosting to Saleor Cloud find it more cost-effective, as our Saleor Cloud is multitenant and   is using shared infrastructure among tens of thousands of projects, which allows us to provide competitive price as well as better user experience by removing the operational burden. It is also important to note that Saleor Cloud is not just hosting but a product on its own, it comes with dashboard and CLI to control data replication, backups, sandboxes, zero-time downtime upgrades, as well as provides SLA, read-replica database and direct support for the Saleor team.

**Comment by traveltek-tmalek** [comment]
This is extremely helpful! Thank you very much :)

**Reply by ashishkpaul** [reply]
Hi @traveltek-tmalek u can also opt oracle cloud servers for free, they are providing always free 2 vps ( 2 virtual core with 1 gb of ram) and one more server with arm64 bit 4 core virtual cpu and 24 gb of ram, u can distribute total 200gb of hdd among with all vps. If u don't want to install on docker,  u can install core api on arm64 with 24gb ram server. core api is very simple to install,  u can Google "how to install python django with gunicorn" and then u have to  install callery and redis server", for dashboard,  compile on local computer and upload build folder to 1 gb ram vps and deploy storefront on 2nd 1gb ram vps

**Reply by traveltek-tmalek** [reply]
Thanks for the information!
